### ISSUE TITLE
ReflectionExtractor::getTypes prioritizes unreliable sources to determine the type

### ISSUE THREAD

**Issue by skobkin** [symfony/symfony#57719]
### Symfony version(s) affected

<= 6.4 =>

Still present in latest stable release.

### Description

`Symfony\Component\PropertyInfo\Extractor\ReflectionExtractor::getTypes` gives more priority to accessor and/or mutator methods over constructor definition even when it uses type hinting.

[`ReflectionExtractor`](https://github.com/symfony/property-info/blob/7.1/Extractor/ReflectionExtractor.php#L143):
```php
public function getTypes(string $class, string $property, array $context = []): ?array
{
    if ($fromMutator = $this->extractFromMutator($class, $property)) {
        return $fromMutator;
    }
    if ($fromAccessor = $this->extractFromAccessor($class, $property)) {
        return $fromAccessor; // data suggesting 'bool' returned before extracting from constructor
    }
    if (
        ($context['enable_constructor_extraction'] ?? $this->enableConstructorExtraction)
        && $fromConstructor = $this->extractFromConstructor($class, $property)
    ) {
        return $fromConstructor;
    }
    if ($fromPropertyDeclaration = $this->extractFromPropertyDeclaration($class, $property)) {
        return $fromPropertyDeclaration;
    }
    return null;
}
```

### How to reproduce

The problem could be reproduced for example when using Symfony Serializer:

We define a class which instantiates it's properties in the constructor and uses type hints there. We also add some method which could be classified as accessor or mutator by `ReflectionExtractor`:

```php
class SomeResponse
{
    public function __construct(
        #[SerializedName('url')]
        public readonly ?string $url, // this type hint is ignored
    ) {
    }

    public function hasUrl(): bool // type is being detected from here
    {
        return !empty($this->url);
    }
}
```

Now we try to deserialize simple JSON:

```php
$serializer->deserialize(SomeResponse::class, '{"url": "http://test"}');
```

We'll get this error in the exception:

```
The type of the "url" attribute for class "SomeResponse" must be one of "bool" ("string" given).
```

### Possible Solution

Maybe check if we can detect the type from constructor and if so, use it as a more reliable source.

**Comment by mtarld** [comment]
Indeed, the `ReflectionExtractor` searches for type in the following order:
1. `setXXX`
1. `isXXX`, `canXXX`, or `hasXXX`
1. `getXXX`
1. `isXXX`, `canXXX`, or `hasXXX` (again)
1. Constructor
1. Property declaration

Therefore, even if your property is public, the result is retrieved from the second step. And I can agree that it is weird.

To improve that, we could update the process to the following:
1. Property declaration **if public**
1. `setXXX`
1. `isXXX`, `canXXX`, or `hasXXX`
1. ...

Or the following:
1. `setXXX`
1. `isXXX`, `canXXX`, or `hasXXX` **if the property is a boolean**
1. `getXXX`
1. `isXXX`, `canXXX`, or `hasXXX`  **if the property is a boolean** (again)
1. Constructor
1. Property declaration

The problem is that these fixes can in some way be considered as BC breaks. So maybe it can be done as a feature. WDYT @dunglas?

In the meantime, @skobkin you can add a useless `setter` to shortcut the process I guess.

**Comment by skobkin** [comment]
@mtarld Thanks, but I worked around this problem as soon as I found out what's the reason. So no help needed. Just came here to report.

I'd settle for the next major release. It's not such a big problem. Just a bit unexpected behavior.

**Comment by alaugks** [comment]
You could submit a proposal as a pull request. Alternatively, you could create and load a custom ReflectionExtractor that would determine the type according to your requirements.

**Comment by alaugks** [comment]
@skobkin You could submit a proposal as a pull request. Alternatively, you could create and load a custom ReflectionExtractor that would determine the type according to your requirements.

**Comment by mtarld** [comment]
I can work on a fix if it is not considered as a BC break. Any thought about that @chalasr?

**Comment by chalasr** [comment]
I'd change that behavior but indeed, doing so would probably have a huge impact so I think it must be done in a BC way.

**Comment by nicolas-grekas** [comment]
Please have a look at #63344
